File: projects/Dot-Dot_Python/File.py
import csv

class File():
    def __init__(self):
        '''=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
        Method               :    __init__
    
        Method parameters    :    self
    
        Method return        :    none
    
        Synopsis             :    This method is a constructor for the file object.
    
        References           :    none
    
        Modifications        :
                              Date            Developer                Notes
                              ----            ---------                -----
                             2023-12-04        V. Arias                Initialization
    
        =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=''' 
        self.file_path = 'assets/highscores.csv' #create a variable to store the file's path
        self.content = [] #declare a list to store the file data
        
    def load_file(self):
        '''=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
        Method               :    load_file
    
        Method parameters    :    self
    
        Method return        :    none
    
        Synopsis             :    This method loads the game information from the file.
    
        References           :    W3Schools (2023) Python Try Except. https://www.w3schools.com/python/python_try_except.asp
                                  Python documentation 3.12.0 (2023) CSV File Reading and Writing. 
                                  https://docs.python.org/3/library/csv.html
                                  
    
        Modifications        :
                              Date            Developer                Notes
                              ----            ---------                -----
                             2023-12-04        V. Arias                Load File
    
        =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=''' 
        try:
            with open(self.file_path, 'r') as file:
                file_reader = csv.reader(file) #read the file
                for row in file_reader:
                    self.content.append(row) #save the row to the content list 
        except:
            print('File Loading Error') # If there is an error, print error message
                                
                
    def save_file(self):
        '''=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
        Method               :    save_file
    
        Method parameters    :    self
    
        Method return        :    none
    
        Synopsis             :    This method saves the game information to the file.
    
        References           :    W3Schools (2023) Python Try Except. https://www.w3schools.com/python/python_try_except.asp
                                  Python documentation 3.12.0 (2023) CSV File Reading and Writing. 
                                  https://docs.python.org/3/library/csv.html
    
        Modifications        :
                              Date            Developer                Notes
                              ----            ---------                -----
                             2023-12-04        V. Arias                Save File
    
        =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=''' 
        try:
            with open(self.file_path, 'w',  newline='') as file: 
                file_writer = csv.writer(file) #intialize file writer
                for row in self.content:
                    file_writer.writerow(row)  #write row to the file
        except:
            print('File Writing Error') # If there is an error, print error message
                
    def validate_name_input(self, input): 
        '''=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
        Method               :    validate_name_input
    
        Method parameters    :    self, input
    
        Method return        :    none
    
        Synopsis             :    This method validates if the name of the user is a valid entry.
    
        References           :    w3schools (2023) Python String isidentifier() Method. 
                                  https://www.w3schools.com/python/ref_string_isidentifier.asp
    
        Modifications        :
                              Date            Developer                Notes
                              ----            ---------                -----
                             2023-12-04        V. Arias                Validate Name Input
    
        =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=''' 
        if not input.isidentifier() or len(input) > 10: #returns true if the string only contains alphanumeric letters (a-z) and (0-9), or underscores (_).
            return True
        return False #return false if string is not valid
    

File: projects/Dot-Dot_C#/File.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.IO;
using System.Linq;
using Microsoft.Xna.Framework.Content;

namespace Dot_Dot
{
    public class File
    {
        private string filePath = "highscores.txt"; //Store the path to the file in a private string
        public List<List<string>> content = new List<List<string>>(); // declare a list of lists to store all the information from the game [name][score]

        public void LoadFile()
        {
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            // Method				:	LoadFile
            //
            // Method parameters	:	none
            //
            // Method return		:	void
            //
            // Synopsis				:   This method reads the contents of the game file. 
            //
            // References			:   Microsoft (2023) StreamWriter Class. https://learn.microsoft.com/en-us/dotnet/api/system.io.streamwriter?view=net-8.0
            //
            //
            // Modifications		:
            //							Date			Developer				Notes
            //							----			---------				-----
            //							2023-12-04		V. Arias				Load File
            //
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            try
            {
                using (StreamReader reader = new StreamReader(filePath)) //declare and initialize the reader object 
                {
                    string line = "";                                    // declare a string to store each line of the file
                    string[] values;                                     // declare a string array to store the values of the lines in the file

                    while ((line = reader.ReadLine()) != null)
                    {
                        values = line.Split(','); // while the next line is not null, load the line and split it on each comma

                        content.Add(values.ToList()); //add the retrieved values to the file content 
                    }
                }
            }
            catch (Exception e)
            {
                return; //If there was an error return
            }
        }

        public void SaveFile()
        {
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            // Method				:	SaveFile
            //
            // Method parameters	:	none
            //
            // Method return		:	void
            //
            // Synopsis				:   This method writes the game data onto the game file.
            //
            // References			:   Microsoft (2023) StreamWriter Class. https://learn.microsoft.com/en-us/dotnet/api/system.io.streamwriter?view=net-8.0
            //
            //
            // Modifications		:
            //							Date			Developer				Notes
            //							----			---------				-----
            //							2023-12-04		V. Arias				Save File
            //
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            try
            {
                using (StreamWriter writer = new StreamWriter(filePath)) //declare and initialize the writer object 
                {
                    foreach (List<string> row in content)
                    {
                        writer.WriteLine(string.Join(",", row)); //write each of the rows in the content list and append them with commas
                    }
                }
            }
            catch(Exception e) 
            { 
                return; //If there was an error return
            }

        }

        public bool ValidateNameInput(string input) //NOT WORKING :(
        {
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            // Method				:	LoadFile
            //
            // Method parameters	:	none
            //
            // Method return		:	void
            //
            // Synopsis				:   This method is a constructor for the game settings and mouse settings. 
            //
            // References			:   none
            //                          
            //
            // Modifications		:
            //							Date			Developer				Notes
            //							----			---------				-----
            //							2023-12-04		V. Arias				LoadFile
            //
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=


            if (!string.IsNullOrWhiteSpace(input) && input.Length < 10) //check that the input is max 10 characters, and not empty or white space
            { 
                for (byte counter = 0; counter < input.Length; counter++)
                {
                    try
                    {
                        char currentCharacter = Char.Parse(input[counter].ToString());  //store every character in the input string one by one
                        if (!Char.IsLetterOrDigit(currentCharacter)) 
                        {
                            return true; // if the character is not a letter or digit, return true
                        }
                    }
                    catch (Exception e)
                    {
                        return true; // if there is an error reading or parsing the character, return true
                    }
                }
                return false; //if no characters caused a return, they are correct, so return false
            }
            return true; //If the string is null, white space or more than 10 characters, return true
        }
    }
}

